lib/core/CDataFrame.cc

```diff
@@ -0,0 +1,513 @@
+}
+
+bool CDataFrame::reserve(std::size_t numberThreads, std::size_t rowCapacity) {
+    if (m_NumberColumns < rowCapacity) {
+        m_RowCapacity = rowCapacity;
+
+        std::size_t stride;
+        std::tie(numberThreads, stride) = this->numberOfThreadsAndStride(numberThreads);
+        LOG_TRACE(<< "numberThreads = " << numberThreads << " stride = " << stride);
+
+        CDataFrameRowSliceReserver reserver{
+            m_NumberColumns, m_RowCapacity - m_NumberColumns};
+
+        std::vector<std::future<bool>> reserves;
+        reserves.reserve(numberThreads);
+        std::size_t j{0};
+        for (std::size_t i = 0; i + 1 < numberThreads; ++i, j += stride) {
+            auto begin = m_Slices.begin() + j;
+            auto end = m_Slices.begin() + j + stride;
+            reserves.push_back(std::async(std::launch::async, reserver, begin, end));
+        }
+        auto begin = m_Slices.begin() + j;
+        auto end = m_Slices.end();
+        if (begin != end) {
+            reserves.push_back(std::async(std::launch::async, reserver, begin, end));
+        }
+
+        bool successful{true};
+        for (auto& reserve : reserves) {
+            successful &= reserve.get();
+        }
+
+        return successful;
+    }
+    return true;
+}
+
+CDataFrame::TReadFuncVecBoolPr CDataFrame::readRows(std::size_t numberThreads,
+                                                    TReadFunc reader) const {
+    if (m_NumberRows == 0) {
+        return {{reader}, true};
+    }
+
+    if (numberThreads == 1) {
+        // This all happens on the main thread to avoid a context switch.
+
+        CDataFrameRowSliceReader sliceReader{reader, m_NumberColumns, m_RowCapacity,
+                                             m_AsyncReadAndWriteToStore};
+        bool successful{sliceReader(m_Slices.begin(), m_Slices.end())};
+        return {{std::move(reader)}, successful};
+    }
+
+    // We use a fixed schedule whereby each reader reads non-overlapping
+    // slices. This means we can get no contention on reads from the slice
+    // vector. This is naturally load balanced because we arrange for each
+    // reader to read, as close as possible, the same number of rows.
+
+    std::size_t stride;
+    std::tie(numberThreads, stride) = this->numberOfThreadsAndStride(numberThreads);
+    LOG_TRACE(<< "numberThreads = " << numberThreads << " stride = " << stride);
+
+    TReadFuncVec readers{numberThreads, reader};
+
+    std::vector<std::future<bool>> reads;
+    reads.reserve(numberThreads);
+    std::size_t j{0};
+    for (std::size_t i = 0; i + 1 < numberThreads; ++i, j += stride) {
+        auto begin = m_Slices.begin() + j;
+        auto end = m_Slices.begin() + j + stride;
+        CDataFrameRowSliceReader sliceReader{readers[i], m_NumberColumns,
+                                             m_RowCapacity, m_AsyncReadAndWriteToStore};
+        reads.push_back(std::async(std::launch::async, sliceReader, begin, end));
```

Comment:
By way of context, I expect there to be fairly few full reads, i.e. there will typically be a lot of work performed for each read. Although if there was a very low memory constraint and we were working on a large data frame then we would typically need 10-100s of reads for a full analysis (this would be a perverse configuration).

Reading around beforehand, Windows seems to have the best implementation of std::async at present (which it does back with a thread pool).

My rationale for not using a thread pool from the outset was:
1) Threading support is getting a lot of attention from the standards committee and I expect support to improve with new std library releases. So if we can get away with std::async at present, we'll get all these benefits for free going forwards.
2) I expect the work performed per read to be non-trivial and I don't have a good intuition for how this will stack up against the cost of any overhead of launching a thread.

That said I agree it is worth profiling and if thread creation proves to be a bottleneck *once we have a full analysis implemented* then we should think about using a thread pool library. Based on the above considerations though, I think I'd prefer to wait and see how the numbers stack up when this work is done. How about for the time being I leave some TODOs in the code to this effect?